src/Helper/HeaderHelper.php

```diff
@@ -17,16 +17,38 @@ public static function getValueAndParameters(string $headerValue): array
         if ($headerValue === '') {
             return [];
         }
-        $parts = preg_split('/\s*;\s*/', $headerValue, -1, PREG_SPLIT_NO_EMPTY);
-        $output = [array_shift($parts)];
-        foreach ($parts as $part) {
-            [$key, $headerValue] = explode('=', $part, 2);
-            if (preg_match('/^"(?<value>.*)"$/', $headerValue, $matches) === 1) {
-                $headerValue = $matches['value'];
-            }
-            $output[$key] = $headerValue;
+        $parts = preg_split('/\s*;\s*/', $headerValue, 2, PREG_SPLIT_NO_EMPTY);
+        $output = [$parts[0]];
+        if (count($parts) === 1) {
+            return $output;
         }
-        return $output;
+        return array_merge($output, self::getParameters($parts[1]));
+    }
+
+    /**
+     * Explode header value to parameters (eg. q=2;version=6)
+     *
+     * @link https://tools.ietf.org/html/rfc7230#section-3.2.6
+     */
+    public static function getParameters(string $headerValue): array
+    {
+        $headerValue = trim($headerValue);
+        if ($headerValue === '') {
+            return [];
+        }
+        $output = new \stdClass();
+        $output->data = [];
+        do {
+            $headerValue = preg_replace_callback(
+                '/^\s*(?<parameter>\w+)\s*=\s*(?:"(?<valueQuoted>[^"]+)"|(?<value>[!#$%&\'*+.^`|~\w-]+))\s*(?:;|$)/',
```

Comment:
- Some parameter names can contain `*` (e.g. `filename*`)
- Quoted string can contain quoted-pair e.g. `\"`,  `\\` [[Basic Rules]](https://www.rfc-editor.org/rfc/rfc2616.html#section-2.2)
- Some headers can contain a listed values in parameters [[Accept]](https://www.rfc-editor.org/rfc/rfc2616.html#section-14.1), [[Accept-Encoding]](https://www.rfc-editor.org/rfc/rfc2616.html#section-14.3):
`Accept: audio/*; q=0.2, audio/basic`
`Accept-Encoding: gzip;q=1.0, identity; q=0.5, *;q=0`

I think it is impossible to solve this problem with one regular expression.
Moreover, some headlines require a personal touch.

Reply:
@roxblnfk is it possible to pull-request a test for it? That would be much easier to fix then.

Reply:
> Some headers can contain a listed values in parameters [Accept], [Accept-Encoding]:

You mean that? https://github.com/yiisoft/yii-web/blob/8881b0421485031976912f5c4cc8321b8dde0468/src/Helper/HeaderHelper.php#L14

> Some parameter names can contain * (e.g. filename*)

in https://tools.ietf.org/html/rfc7230#section-3.2.6:
> Delimiters are chosen from the set of US-ASCII visual characters not allowed in a token (DQUOTE and "(),/:;<=>?@[\]{}").

`*` not allowed, but it's true that the token contains more characters.

> Quoted string can contain quoted-pair e.g. \", \\ [Basic Rules]
https://tools.ietf.org/html/rfc7230#section-3.2.6

agree

Reply:
> You mean that?

Yes sure :-| This is more about parsing a list of header values

Reply:
@roxblnfk this is a function that has been used a lot, would you to improve further with new tests in a new PR?

Reply:
@kamarton i have other tasks now. I can deal with this problem, but if only after some time

Reply:
https://github.com/yiisoft/yii-web/issues/184